### PR TITLE
Use console logs to determine web test results

### LINKED ISSUE #1960: tf-web-test reports success when there is a failure
Repro case:

1. Add below code to tensorboard/plugins/graph/tf_graph_common/test/parser-test.ts
```ts
  it('fails', () => {
    const pbtxt = tf.graph.test.util.stringToArrayBuffer(``);
    return tf.graph.parser.parseGraphPbTxt(pbtxt)
        .then(() => assert.fail('Should NOT resolve'));
  })
```
2. [Optional] Confirm that it fails when run with `bazel run tensorboard/plugins/graph/tf_graph_common/test:test_web_library`
3. Run `bazel test tensorboard/plugins/graph/tf_graph_common/...`
4. Expected failure but it passes.

### PR DESCRIPTION
**Pull request by erzel**
Fixes #1960.

### CODE REVIEW

File: tensorboard/functionaltests/wct_test_driver.py [hunk 1]
```diff
@@ -47,6 +50,30 @@ def create_test_class(binary_path, web_path):
...
+        self.passed = True
+        return True
+      # Here, we still don't know.
+      return False
+
+    def _log_contains(self, messages, text):
+      for message in messages:
+        if text in message:
+          # Print log message as an aid for debugging.
+          print("Found: %s in log entry message: %s" % (text, message))
```
Comment (wchargin) [inline tensorboard/functionaltests/wct_test_driver.py]: Note that the message is just `"%c1 failing test"`, so I’m not sure how
much this actually helps debugging (e.g., it doesn’t give you the name
of the failing test case), but this is fine with me.

File: tensorboard/functionaltests/wct_test_driver.py [hunk 2]
```diff
@@ -90,13 +117,14 @@ def tearDown(cls):
...
       url = "http://localhost:%s%s" % (self.port, web_path)
       driver.get(url)
-      wait.WebDriverWait(driver, 10).until(
-          expected_conditions.title_contains("test"))
-      title = driver.title
-      if "failing test" in title or "passing test" not in title:
-        self.fail(title)
+      browser_log_indicates_result = BrowserLogIndicatesResult()
+      wait.WebDriverWait(driver, 10).until(browser_log_indicates_result)
+      if not browser_log_indicates_result.passed:
```
Comment (wchargin) [inline tensorboard/functionaltests/wct_test_driver.py]: `self.assertTrue(browser_log_indicates_result.passed)` is simpler, yes?

File: tensorboard/functionaltests/wct_test_driver.py [hunk 3]
```diff
@@ -47,6 +50,30 @@ def create_test_class(binary_path, web_path):
...
+    def __init__(self):
+      self.passed = False
+    def __call__(self, driver):
+      # Scan through the log entries and search for a text indicating whether
+      # the test passed or failed.
+      messages = [entry['message'] for entry in driver.get_log('browser')]
+      if self._log_contains(messages, 'failing test'):
+        self.passed = False
+        return True
+      if self._log_contains(messages, 'test suite passed'):
```
Comment (wchargin) [inline tensorboard/functionaltests/wct_test_driver.py]: Do you have a reference that indicates that the driver will only print
“test suite passed” once _all_ tests have passed? For instance, it’s not
obvious to me that this waits for all top-level `describe` blocks—I
checked, and it appears to, but I don’t see any guarantee of that.

File: tensorboard/functionaltests/wct_test_driver.py (hunk 2 above)
Comment (erzel) [reply tensorboard/functionaltests/wct_test_driver.py]: I had that initially, but then the error message for a passing test is something like 'False' is 'True', which I didn't like so much. It's only slightly better with 'fail'.

File: tensorboard/functionaltests/wct_test_driver.py (hunk 3 above)
Comment (erzel) [reply tensorboard/functionaltests/wct_test_driver.py]: I'm not a 100% certain, but if I look at console.js in chrome's debugger there's a function called logSummary that seems to be called only at the end of the mocha run that logs 'test suite passed'.

File: tensorboard/functionaltests/wct_test_driver.py (hunk 1 above)
Comment (erzel) [reply tensorboard/functionaltests/wct_test_driver.py]: Good point. Changed to output entire log at end of test.

**Comment by erzel** [comment]
Thanks @wchargin !
PTAL.

File: tensorboard/functionaltests/wct_test_driver.py (hunk 3 above)
Comment (wchargin) [reply tensorboard/functionaltests/wct_test_driver.py]: Okay, no problem—if we find out that this is insufficient then we can
update this test code again. :-)

File: tensorboard/functionaltests/wct_test_driver.py (hunk 2 above)
Comment (wchargin) [reply tensorboard/functionaltests/wct_test_driver.py]: Got it; thanks.

File: tensorboard/functionaltests/wct_test_driver.py [hunk 4]
```diff
@@ -90,13 +121,26 @@ def tearDown(cls):
...
+      try:
+        wait.WebDriverWait(driver, 10).until(browser_log_indicates_result)
+        if not browser_log_indicates_result.passed:
+          self.fail()
+      finally:
+        # Print log as an aid for debugging.
+        log = browser_log_indicates_result.log + driver.get_log("browser")
+        self._printLog(log)
+
+    def _printLog(self, entries):
```
Comment (wchargin) [inline tensorboard/functionaltests/wct_test_driver.py]: `_print_log`, not `_print_log`

File: tensorboard/functionaltests/wct_test_driver.py [hunk 5]
```diff
@@ -90,13 +121,26 @@ def tearDown(cls):
...
+      print ("Browser log follows:")
```
Comment (wchargin) [inline tensorboard/functionaltests/wct_test_driver.py]: remove space between `print` and parenthesis

File: tensorboard/functionaltests/wct_test_driver.py [hunk 6]
```diff
@@ -90,13 +121,26 @@ def tearDown(cls):
...
-        self.fail(title)
+      browser_log_indicates_result = BrowserLogIndicatesResult()
+      try:
+        wait.WebDriverWait(driver, 10).until(browser_log_indicates_result)
+        if not browser_log_indicates_result.passed:
+          self.fail()
+      finally:
+        # Print log as an aid for debugging.
+        log = browser_log_indicates_result.log + driver.get_log("browser")
+        self._printLog(log)
```
Comment (wchargin) [inline tensorboard/functionaltests/wct_test_driver.py]: Nice; thanks! It looks like this prints the assertion failures
(“expected 0, got 1”), which is definitely better than the previous
behavior even though the test case name isn’t logged.

File: tensorboard/functionaltests/wct_test_driver.py (hunk 5 above)
Comment (erzel) [reply tensorboard/functionaltests/wct_test_driver.py]: Done.

File: tensorboard/functionaltests/wct_test_driver.py (hunk 4 above)
Comment (erzel) [reply tensorboard/functionaltests/wct_test_driver.py]: Oops. Done.

File: tensorboard/functionaltests/wct_test_driver.py (hunk 6 above)
Comment (erzel) [reply tensorboard/functionaltests/wct_test_driver.py]: Yep. It would be nice to try to get that as well, but I'll stop here for now.
PTAL. Thanks!

Comment (erzel) [reply tensorboard/functionaltests/wct_test_driver.py]: Note that I don't have write-access and cannot merge.

**Pull request merged**